src/builtin.cpp

```diff
@@ -1777,31 +1783,64 @@ static int builtin_random(parser_t &parser, io_streams_t &streams, wchar_t **arg
             }
         }
     }
-
+    // argument parsing
+    bool parse_error = false;
+    auto parse_ll = [&](const wchar_t* str) {
+        long long ll = fish_wcstoll(str);
+        if (errno) {
+            streams.err.append_format(L"%ls: %ls is not a valid integer\n", argv[0], str);
+            parse_error = true;
+        }
+        return ll;
+    };
     int arg_count = argc - w.woptind;
+    long long start, end, step;
     if (arg_count == 0) {
-        long res;
-        if (!seeded) {
-            seeded = 1;
-            srand48_r(time(0), &seed_buffer);
-        }
-        lrand48_r(&seed_buffer, &res);
-        streams.out.append_format(L"%ld\n", res % 32768);
+        start = 0;
+        end = 32767;
+        step = 1;
     } else if (arg_count == 1) {
-        long foo = fish_wcstol(argv[w.woptind]);
-        if (errno) {
-            streams.err.append_format(_(L"%ls: Seed value '%ls' is not a valid number\n"), argv[0],
-                                      argv[w.woptind]);
+        long long seed = parse_ll(argv[w.woptind]);
+        if (!parse_error) {
+            engine.seed(seed);
+            return STATUS_BUILTIN_OK;
+        } else {
             return STATUS_BUILTIN_ERROR;
         }
-        seeded = 1;
-        srand48_r(foo, &seed_buffer);
+    } else if (arg_count == 2) {
+        start = parse_ll(argv[w.woptind]);
+        step = 1;
+        end = parse_ll(argv[w.woptind + 1]);
+    } else if (arg_count == 3) {
+        start = parse_ll(argv[w.woptind]);
+        step = parse_ll(argv[w.woptind + 1]);
+        end = parse_ll(argv[w.woptind + 2]);
     } else {
-        streams.err.append_format(_(L"%ls: Expected zero or one argument, got %d\n"), argv[0],
-                                  argc - w.woptind);
-        builtin_print_help(parser, streams, argv[0], streams.err);
+        streams.err.append_format(BUILTIN_ERR_TOO_MANY_ARGUMENTS, argv[0]);
         return STATUS_BUILTIN_ERROR;
     }
+
+    // error handling
+    if (parse_error) {
+        return STATUS_BUILTIN_ERROR;
+    } else if (start > end) {
+        streams.err.append_format(L"%ls: END must be greater than or equal to START\n", argv[0]);
+        return STATUS_BUILTIN_ERROR;
+    } else if (step <= 0) {
+        streams.err.append_format(L"%ls: STEP must be a positive integer\n", argv[0]);
+        return STATUS_BUILTIN_ERROR;
+    }
+
+    // rng-ing
+    long long result;
+    if (end - start < step) {
+        // nine nine nine nine nine nine
+        result = start;
```

Comment:
I'd hate to lose the dilbert reference, but I'm not sure returning something deterministic is the right thing to do. Error?